app/src/main/java/unipd/se18/ocrcamera/PreProcessing.java

```diff
@@ -0,0 +1,369 @@
+     * @param image The image you want to analyze
+     * @return the angle between the text and the horizontal. 0 if it fails
+     * @author Thomas Porro (g1)
+     */
+    private double computeSkew(Bitmap image){
+
+        //Turns the image into a matrix
+        Mat img;
+        try{
+            img = IPUtils.conversionBitmapToMat(image);
+        } catch (ConversionFailedException error){
+            Log.e(TAG, error.getErrorMessage());
+            return 0;
+        }
+
+
+        Mat grayscale = new Mat();
+        Imgproc.cvtColor(img, grayscale, Imgproc.COLOR_RGB2GRAY);
+       /*
+            Method used for debug
+            IPUtils.save(img, "grayScale", ".jpg");
+        */
+
+        //Invert the colors of "img" onto itself
+        Core.bitwise_not(img, img);
+
+        //Detect the edges in the image
+        Mat canny = IPBuilder.doCanny(new IPBuilder.CannyBuilder(grayscale)
+                .withMinThreshold(50)
+                .withMaxThreshold(200)
+                .withApertureSize(3)
+                .withL2gradient(false));
+
+        //Create a 4 dimensions vector using matrix
+        MatOfInt4 lines = new MatOfInt4();
+
+        //Process the image with the Probabilistic Hough Transform
+        double rho = 1;
+        double theta = Math.PI / 180;
+        int threshold = 50;
+        double minLineLenght = 50;
+        double maxLineGap = 10;
+        Imgproc.HoughLinesP(grayscale, lines, rho, theta, threshold, minLineLenght, maxLineGap);
+
+        double meanAngle = 0;
+        Log.d(TAG, "rows = " + lines.cols() + "\ncols = " + lines.cols());
+
+        //Analizes the text line per line
+        for (int i = 0; i < lines.rows(); i++) {
+            //Get points from the beginning and the ending of the line of text
+            double[] vec = lines.get(i, 0);
+
+            //First point
+            double x1 = vec[0];
+            double y1 = vec[1];
+
+            //Second point
+            double x2 = vec[2];
+            double y2 = vec[3];
+
+            //Sum all the angle of the lines of text
+            meanAngle += Math.atan2(y2 - y1, x2 - x1);
+        }
+
+        //Calculate the meanAngle by dividing it with the number of rows
+        meanAngle /= lines.rows();
+
+        //Transform the angle in degrees
+        double degreesAngle = Math.toDegrees(meanAngle);
+        Log.i(TAG, "Mean angle=" + degreesAngle);
+        return degreesAngle;
+    }
+
+    /**
+     * Rotate the given bitmap with the given angle
+     * @param original The image that we want to corrected
+     * @param degrees The degrees of the angle we want to rotate
+     * @return the rotatedImage. If the image is null return the original image
+     * @author Thomas Porro(g1), Giovanni Fasan (g1), Oscar Garrido (g1)
+     */
+    private static Bitmap rotateBitmap(Bitmap original, double degrees) {
+
+        //TODO not correct because it returns null, maybe we should add a @NonNull
```

Comment:
What is this todo? Have I to review this method?

Reply:
The idea is to later add the impossibility of passing a null value as a parameter